src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/PopupControlService.cs

```diff
@@ -236,557 +221,557 @@ private void ProcessKeyUp(object sender, KeyEventArgs e)
-                        _lastObjectWithToolTip = null;
-                    }
-                }
+            get { return _lastMouseToolTipOwner.GetValue(); }
+            set { _lastMouseToolTipOwner.SetValue(value); }
+        }
 
-                return null;
-            }
+        private DependencyObject GetOwner(ToolTip t)
+        {
+            return t?.GetValue(OwnerProperty) as DependencyObject;
+        }
 
-            set
+        // a pending request is represented by a sentinel ToolTip object that carries
+        // the owner and the trigger action (only).  There's never more than one
+        // pending request, so we reuse the same sentinel object.
+        private ToolTip SentinelToolTip(DependencyObject o, ToolTipService.TriggerAction triggerAction)
+        {
+            // lazy creation, because we cannot create it in the ctor (infinite loop with FrameworkServices..ctor)
+            if (_sentinelToolTip == null)
             {
-                if (value == null)
-                {
-                    _lastObjectWithToolTip = null;
-                }
-                else if (_lastObjectWithToolTip == null)
-                {
-                    _lastObjectWithToolTip = new WeakReference(value);
-                }
-                else
-                {
-                    _lastObjectWithToolTip.Target = value;
-                }
+                _sentinelToolTip = new ToolTip();
             }
+
+            _sentinelToolTip.SetValue(OwnerProperty, o);
+            _sentinelToolTip.FromKeyboard = ToolTipService.IsFromKeyboard(triggerAction);
+            return _sentinelToolTip;
         }
 
-        private DependencyObject LastChecked
+        #region Safe Area
+
+        private void SetSafeArea(ToolTip tooltip)
         {
-            get
+            SafeArea = null;     // default is no safe area
+
+            // safe area is only needed for tooltips triggered by mouse
+            if (tooltip != null && !tooltip.FromKeyboard)
             {
-                if (_lastChecked != null)
+                UIElement owner = GetOwner(tooltip) as UIElement;
+                PresentationSource presentationSource = (owner == null) ? null : PresentationSource.CriticalFromVisual(owner);
+                if (presentationSource != null)
                 {
-                    DependencyObject o = (DependencyObject)_lastChecked.Target;
-                    if (o != null)
-                    {
-                        return o;
-                    }
-                    else
+                    // get owner and tooltip rectangles, in owner window's client coords
+                    Rect rectElement = new Rect(new Point(0, 0), owner.RenderSize);
+                    Rect rectRoot = PointUtil.ElementToRoot(rectElement, owner, presentationSource);
+                    Rect ownerRect = PointUtil.RootToClient(rectRoot, presentationSource);
+
+                    Rect screenRect = tooltip.GetScreenRect();
```

Comment:
Can it work well when we have two Moniter with difference DPI and the owner displayed in the Moniter one and the tooltip displayed in the Moniter two?

Reply:
Tooltips are always displayed on the same monitor as the owner.